platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/PreConsensusEventFileManager.java

```diff
@@ -0,0 +1,463 @@
+        if (!Files.exists(databaseDirectory)) {
+            Files.createDirectories(databaseDirectory);
+        }
+
+        files = new RandomAccessDeque<>(INITIAL_RING_BUFFER_SIZE);
+
+        try (final Stream<Path> fileStream = Files.walk(databaseDirectory)) {
+            fileStream
+                    .filter(f -> !Files.isDirectory(f))
+                    .map(PreConsensusEventFileManager::parseFile)
+                    .filter(Objects::nonNull)
+                    .sorted()
+                    .forEachOrdered(buildFileHandler(configuration.permitGaps()));
+        }
+
+        // Measure the size of each file.
+        for (final PreConsensusEventFile file : files) {
+            totalFileByteCount += Files.size(file.path());
+        }
+
+        if (files.size() > 0) {
+            metrics.getPreconsensusEventFileOldestGeneration()
+                    .set(files.getFirst().minimumGeneration());
+            metrics.getPreconsensusEventFileYoungestGeneration()
+                    .set(files.getLast().maximumGeneration());
+            final Duration age = Duration.between(time.now(), files.getFirst().timestamp());
+            metrics.getPreconsensusEventFileOldestSeconds().set(age.toSeconds());
+        }
+        updateFileSizeMetrics();
+    }
+
+    /**
+     * Parse a file into a PreConsensusEventFile wrapper object.
+     *
+     * @param path the path to the file
+     * @return the wrapper object, or null if the file can't be parsed
+     */
+    private static PreConsensusEventFile parseFile(final Path path) {
+        try {
+            return PreConsensusEventFile.of(path);
+        } catch (final IOException exception) {
+            // ignore any file that can't be parsed
+            logger.warn(EXCEPTION.getMarker(), "Failed to parse file: {}", path, exception);
+            return null;
+        }
+    }
+
+    /**
+     * Build a handler for new files parsed from disk. Does basic sanity checks on the files, and adds them to the file
+     * list if they are valid.
+     *
+     * @param permitGaps if gaps are permitted in sequence number
+     * @return the handler
+     */
+    private Consumer<PreConsensusEventFile> buildFileHandler(final boolean permitGaps) {
+
+        final ValueReference<Long> previousSequenceNumber = new ValueReference<>(-1L);
+        final ValueReference<Long> previousMinimumGeneration = new ValueReference<>(-1L);
+        final ValueReference<Long> previousMaximumGeneration = new ValueReference<>(-1L);
+        final ValueReference<Instant> previousTimestamp = new ValueReference<>();
+
+        return descriptor -> {
+            fileSanityChecks(
+                    permitGaps,
+                    previousSequenceNumber,
+                    previousMinimumGeneration,
+                    previousMaximumGeneration,
+                    previousTimestamp,
+                    descriptor);
+
+            previousSequenceNumber.setValue(descriptor.sequenceNumber());
```

Comment:
The file sanity checks make sure that the previous sequence number is one less than the current sequence number. If I set this value before the sanity checks, it would observe the "current" to be the same as the "previous".